src/main/java/com/cloudbees/dockerpublish/DockerBuilder.java

```diff
@@ -49,24 +51,25 @@
 
     private static final Pattern IMAGE_BUILT_PATTERN = Pattern.compile("Successfully built ([0-9a-f]+)");
 
-    private final String repoName;
-    private final boolean noCache;
-    private final boolean forcePull;
-    private final String dockerfilePath;
-    private final boolean skipBuild;
-    private final boolean skipDecorate;
+    @CheckForNull
+    private final DockerServerEndpoint server;
+    @CheckForNull
+    private DockerRegistryEndpoint registry;
```

Comment:
I see no reason for either of these to be `@CheckForNull`. They are always nonnull when new instances are constructed via databinding. When upgrading, you would just need a `readResolve` (here in the builder, unrelated to the descriptor’s `readResolve`) to set them to blank values.